docs/design.md

```diff
@@ -0,0 +1,508 @@
+* If the external resource does exists, but is not update to date, it will call `Update()` to delegate the actual resource updating job to provider. Once it's done successfully, it will requeue to wait for the next run of `Observe()` controlled by poll interval to discover any state difference between the external resource and the managed resource.
+* When the managed resource is deleted, it will call `Delete()` that allows provider to do clean up job for the external resource. After then, just as `Create()/Update()` does, it will requeue to wait for the next run of `Observe()` so that it can keep synchronizing the state between the external resource and the managed resource.
+* For all above CRUD methods, if there is an error occurred, it will report the error and requeue to wait for the next run of `Observe()` to give it another try. This is the typical Kubernetes reconciliation logic.
+
+### Mapping Ansible Run to Managed Resource Lifecycle
+
+In Ansible provider, `AnsibleRun` is the managed resource which is used to define the desired state on target system. It leverages the cooperation between Managed Resource Reconciler and Ansible provider to ensure the consistency between the desired state and the actual state. In order to support this, we need to define how `AnsibleRun` can be managed by Managed Resource Reconciler to follow the Managed Resource Lifecyle.
+
+### Preparing Ansible Contents
+
+To support Ansible run, Ansible provider requires the Ansible contents to be available before run. This may involve the work to retrieve Ansible contents from a remote place or to create a `playbook.yml` file including the inline content, then store into the local working directory.
+
+In Ansible provider, this is supported by implement the above logic in `Connect()`.
+Once an `AnsibleRun` resource is created, the reconciler will call the provider method `Connect()` to retrieve Ansible contents from the remote or generate inline playbook file which depends on how we define `AnsibleRun`.
+
+### Ansible Run Policy
+
+Once Ansible contents are available, we can start the Ansible run. Ansible provider supports a couple of run policies to fulfill different types of requirements. The policy is represented as annotation `ansible.crossplane.io/runPolicy`, which can be applied to `AnsibleRun` resource, to instruct the provider how to run the corresponding Ansible contents.
+
+#### Policy ObserveAndDelete
+
+This is the default policy and probably the most commonly used policy to manage Ansible run. When this policy is applied, the provider uses `Observe()` to handle the case when the managed resource `AnsibleRun` is present, and uses `Delete()` to handle the case when the resource is absent. Both `Observe()` and `Delete()` will call the same set of Ansible contents.
+
```

Comment:
`and uses Delete() to handle the case when the resource is absent` are  you talking about managed resource here ?

Reply:
`the provider uses Observe() to handle the case when the managed resource AnsibleRun is present, and uses Delete() to handle the case when the resource is absent` ==> it is not the native flow of reconciliation defined by the crossplane runtime  https://github.com/multicloudlab/crossplane-provider-ansible/pull/27/files#diff-b0929dacd4f1fc192cd308c1a9db2e85154ea742ccec315518e659d7f6a3efc7R376

Reply:
> Are you talking about managed resource here ?
Yes. Maybe I should make it clearer here. Thanks.

> it is not the native flow of reconciliation defined by the crossplane runtime
Did not quite understand what you mean here. Could you elaborate it a bit more? It is something like a sub-set of the whole reconciliation flow defined by crossplane runtime. Originally, I thought we should strictly map Ansible run to the crossplane reconciliation flow (i.e.: Connect + CRUD). However, after I checked Ansible Operator and the usual way of how people run Ansible in Ansible community, I think that to map a sub-set might be a much realistic option at least for the first few releases of Ansible provider.

Reply:
hmm ok so it is a sub-set of the reconciliation flow defined by crossplane runtime.  I think I can't see how to emerge this sub-set  in the crossplane reconciliation flow. I had the impression that you rewrite the crossplane flow ( more specificly the `Reconcile` func itself ) ; when you said `uses Observe() to handle the case when the managed resource AnsibleRun is present, and uses Delete() to handle the case when the resource is absent`

Reply:
We do not rewrite the crossplane lifecycle flow, we just simplify it by binding Ansible run to some of its phases. Essentially, crossplane lifecycle flow is composed with a set of phases (or methods):
- connect
- observe
- create
- update
- delete 

On the other side, for an Ansible role or playbook in our case, it is typically used to provision e.g.: a service on GCP or an app on VM.

When I say `uses Observe() to handle the case when the managed resource AnsibleRun is present`, it means, when user creates an `AnsibleRun` CR (a.k.a. the managed resource in Crossplane's term), it means he/she claims to request the service on GCP or the app on VM, this will trigger the Ansible role or playbook to provision it for the user. To trigger this, we need to map the call made to Ansible in a certain crossplane lifecycle phase, and `observe` is the most appropriate one both from concept and implementation perspective (it runs periodically). With that, we will call the Ansible role or playbook in `Observe()` method and pass `state: present` into it.

When I say `uses Delete() to handle the case when the resource is absent`, it means, when user deletes the `AnsibleRun` CR, it means he/she claims do not need the service on GCP or the app on VM any more, this will trigger the same Ansible role or playbook to delete it for the user. To trigger this, again, we map the call to Ansible to `delete` phase. With that, we will call the Ansible role or playbook in `Delete()` method and pass `state:absent` into it.

We did not touch `create` and `update` in this case, as that requires to differentiate when to create/update, and when not, this is covered by the other policy, which is run in check mode in `Observe()`.

So, this looks like a simplified/collapsed version of the complete crossplane lifecycle flow.

Does it make sense?

Reply:
OK got it! 
**in more generic version:** the ansible provider controller delegates the action to `ansible` binary to make changes in the remote resource in the cloud. This is the major difference compared to the other crossplane provider (particularly terraform).  We no longer make direct API calls to the cloud using local binaries or golang libraries but we rely on the local Ansible content to make these calls/changes
I think we also need a scheme for this :). But this can be on following PR 
<img width="876" alt="Capture d’écran 2022-04-18 à 12 11 26" src="https://user-images.githubusercontent.com/11649303/163794662-c9372624-48ed-4335-ba2f-4053ac307875.png">

Reply:
yeah, we are on the same page now! 👍